src/libutil/url.cc

```diff
@@ -408,21 +408,107 @@ ParsedUrlScheme parseUrlScheme(std::string_view scheme)
+ * https://git-scm.com/book/ms/v2/Git-on-the-Server-The-Protocols
+ *
+ * A good reference is libgit2 also allows scp style
+ * https://github.com/libgit2/libgit2/blob/58d9363f02f1fa39e46d49b604f27008e75b72f2/src/util/net.c#L806
+ */
+static std::optional<ScpLike> parse_scp(std::string_view s) noexcept
+{
+    if (s.empty() || s.front() == '/')
+        return std::nullopt;
+
+    // Find the colon that separates host from path.
+    // Find the right-most since ipv6 has colons
+    const auto colon = s.rfind(':');
+    if (colon == std::string_view::npos)
+        return std::nullopt;
+
+    // Split head:[path]
+    const auto head = s.substr(0, colon);
+    const auto path = s.substr(colon + 1);
+
+    if (head.empty())
+        return std::nullopt;
+
+    ScpLike out{};
+
+    // Optional user@
+    const auto at = head.find('@');
+    std::string_view host = head;
+
+    if (at != std::string_view::npos) {
+        // empty user
+        if (at == 0)
+            return std::nullopt;
+        out.user = head.substr(0, at);
+        host = head.substr(at + 1);
+        // empty host
+        if (host.empty())
+            return std::nullopt;
+    }
+
+    // Host can be bracketed IPv6: [....]
+    if (host.front() == '[') {
+        // Must contain a closing ']' and nothing between ']' and ':'
+        const auto rb = host.find(']');
+        if (rb == std::string_view::npos)
+            return std::nullopt;
+        out.host = host.substr(0, rb + 1);
+    } else {
+        out.host = host;
+    }
+
+    out.path = path;
+    return out;
+}
+
 ParsedURL fixGitURL(const std::string & url)
 {
-    std::regex scpRegex("([^/]*)@(.*):(.*)");
-    if (!hasPrefix(url, "/") && std::regex_match(url, scpRegex))
-        return parseURL(std::regex_replace(url, scpRegex, "ssh://$1@$2/$3"));
+    if (auto r = boost::urls::parse_uri(url); r && r->has_scheme() && r->has_authority()) {
+        return parseURL(url);
+    }
+
     if (hasPrefix(url, "file:"))
         return parseURL(url);
-    if (url.find("://") == std::string::npos) {
-        return ParsedURL{
-            .scheme = "file",
-            .authority = ParsedURL::Authority{},
-            .path = splitString<std::vector<std::string>>(url, "/"),
-        };
+
+    if (auto scp = parse_scp(url)) {
+        auto [user, host, path] = *scp;
+        std::string out = "ssh://";
+        if (!user.empty()) {
+            out.append(user);
+            out.push_back('@');
+        }
+        out.append(host);
+        if (!path.empty()) {
+            // Only absolute paths are allowed for ssh variant
+            if (!hasPrefix(path, "/")) {
+                out.push_back('/');
+            }
+            out.append(path);
+        }
+        return parseURL(out);
+    }
+
+    // if the url does not start with forward slash, add one
+    auto modifiedUrl = url;
+    if (!hasPrefix(url, "/")) {
+        modifiedUrl = "/" + url;
     }
```

Comment:
This is new from the merge; without it it actually fails to create a ParsedURL since it's relative but we are providing an authority.

The other option is to return broken URLs reflecting the relative path... which we already do above if they provide the scheme `file://` -- there is a test case that shows this.